File: src/Ast.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const MultiArrayList = std.MultiArrayList;
const Parser = @import("Parser.zig");
const tokenizer = @import("tokenize.zig");
const Token = tokenizer.Token;
const Tokenizer = tokenizer.Tokenizer;
const Ast = @This();

source: [:0]const u8,
nodes: Node.List.Slice,
extra_data: []Node.Idx,
tokens: []const Token,

pub const TokenIdx = u32;

pub const Node = struct {
    tag: Tag,
    main_idx: TokenIdx,
    data: Data,

    pub const Idx = enum(u32) { _ };
    // 0 is reserved to refer to the root node, so 0 may be used as `null` index;
    pub const null_node: Idx = @enumFromInt(0);

    pub const Tag = enum {
        // `main_idx` is always 0
        // data is list of decls as extra_data[lhs..rhs]
        root,
        // `pub? (const|var) ident(: type_expr) = expr;`
        // `main_idx` is (const|var)
        // lhs is optional `type_expr`, indexes extra_data
        // rhs is `expr`, indexes nodes
        var_decl,
        // `struct { fields* }`
        // `main_idx` is {
        // TODO: This needs to be revisted wrt root
        // If there are no fields or decls, then data is 0
        // If there is a single field or decl, then lhs is 0 and data.rhs is a single field or decl indexing nodes
        // If there are 2 or more fields or decls, then data.lhs is extra_index index of a `StructBody` and data.rhs is 0
        struct_decl,
        // `module { module_fields* module_statements* }`
        // `main_idx` is module
        // `lhs` is extra_index index of the start of a list of node indexes
        // `rhs` is extra_index index of the end of a list of node indexes
        // `lhs`..`rhs` can contain `container_field`s or `comb_def`s
        module_decl,
        // `pub? (comb|fn) ident(args*) ret_ty { statements* }`
        // `main_idx` is `ident`
        // `lhs` is index of `subroutine_sig`
        // `rhs` is index of `subroutine_body`
        subroutine_decl,
        // `(args*) ret_ty`
        // `main_idx` is `(`
        // `lhs` is extra_index index of `SubroutineArgs` which indexes into `extra_data`
        // `rhs` is index of ret_ty
        subroutine_sig,
        // `{ statements* }`
        // `main_idx` is `{`
        // `lhs` is the start of extra_index index of subroutine statements
        // `rhs` is the end of extra_index indexes of subroutine statmenets
        subroutine_body,
        // `ident: type_expr`
        // `main_idx` is ident
        // lhs is `nodes` index of `type_expr`
        // rhs is unused
        subroutine_arg,
        // `ident: type_expr( = expr)?`
        // `main_idx` is ident
        // lhs is `type_expr`, indexes `nodes`
        // rhs is optional `expr`, indexes `nodes`
        container_field,
        // `lhs or rhs`
        // `main_idx` is `or`
        @"or",
        // `lhs and rhs`
        // `main_idx` is `and`
        @"and",
        // `lhs < rhs`
        // `main_idx` is <
        lt,
        // `lhs > rhs`
        // `main_idx` is >
        gt,
        // `lhs <= rhs`
        // `main_idx` is <=
        lte,
        // `lhs >= rhs`
        // `main_idx` is >=
        gte,
        // `lhs == rhs`
        // `main_idx` is ==
        eq,
        // `lhs != rhs`
        // `main_idx` is !=
        neq,
        // `lhs & rhs`
        // `main_idx` is &
        bit_and,
        // `lhs | rhs`
        // `main_idx` is |
        bit_or,
        // `lhs ^ rhs`
        // `main_idx` is ^
        bit_xor,
        // `lhs + rhs`
        // `main_idx` is `+`
        add,
        // `lhs - rhs`
        // `main_idx` is `-`
        sub,
        // `lhs * rhs`
        // `main_idx` is `*`
        mul,
        // `lhs / rhs`
        // `main_idx` is `/`
        div,
        // `&var? lhs`
        // `main_idx` is &
        // `lhs` is an expression
        // `rhs` is unused
        reference,
        // `*var? lhs`
        // `main_idx` is *
        // `lhs` is an expression
        // `rhs` is unused
        ptr_ty,
        // `lhs assignment_op rhs`
        // `main_idx` is assignment_op
        // `lhs` is an lvalue expression
        // `rhs` is an expression
        assignment,
        // `lhs.rhs`
        // `main_idx` is `.`
        // `lhs` is an identifier
        // `rhs` is unused
        member,
        // `ident`
        // `main_idx` is `ident`
        // `lhs` and `rhs` unused
        identifier,
        // `int`
        // `main_idx` is `int`
        // `lhs` and `rhs` unused
        int,
        // `return e`
        // `main_idx` is `return`
        // `lhs` is `node` index of `e`
        // `rhs` is unused
        @"return",
    };

    // TODO change this to be bare union between Idx and ExtraIdx for debug type safety
    pub const Data = struct {
        lhs: Idx,
        rhs: Idx,
    };

    pub const SubList = struct {
        start: Idx,
        end: Idx,
    };

    pub const StructBody = struct {
        fields_start: Idx,
        fields_end: Idx,
        decls_start: Idx,
        decls_end: Idx,
    };

    // Indexes into extra data
    // Each index is a `comb_arg`
    pub const CombArgs = struct {
        args_start: Idx,
        args_end: Idx,
    };

    pub const List = MultiArrayList(Node);
};

pub fn parse(allocator: Allocator, source: [:0]const u8) !Ast {
    var tokens = try std.ArrayList(Token).initCapacity(allocator, source.len);
    var t = Tokenizer.init(source);

    while (true) {
        const token = t.next();
        try tokens.append(token);
        if (token.tag == .invalid) @panic("Uh oh invalid tag");
        if (token.tag == .eof) break;
    }

    var parser = Parser{
        .allocator = allocator,
        .source = source,
        .tokens = try tokens.toOwnedSlice(),
    };
    defer parser.deinit();
    try parser.nodes.setCapacity(allocator, parser.tokens.len);
    try parser.parseRoot();

    return Ast{
        .source = source,
        .nodes = parser.nodes.toOwnedSlice(),
        .extra_data = try parser.extra_data.toOwnedSlice(allocator),
        .tokens = parser.tokens,
    };
}

pub fn deinit(self: *Ast, allocator: Allocator) void {
    self.nodes.deinit(allocator);
    allocator.free(self.extra_data);
    allocator.free(self.tokens);
}

// Useful types for constructing assembled information about a Node
pub const Assembled = struct {
    pub const Struct = struct {
        token: TokenIdx,
        members: []const Node.Idx,
    };

    pub const VarDecl = struct {
        token: TokenIdx,
        type_expr: Node.Idx,
        expr: Node.Idx,
    };
};

pub fn assembledStruct(self: *const Ast, node_idx: Node.Idx) ?Assembled.Struct {
    const idx = @intFromEnum(node_idx);
    const lhs = @intFromEnum(self.nodes.items(.data)[idx].lhs);
    const rhs = @intFromEnum(self.nodes.items(.data)[idx].rhs);
    return switch (self.nodes.items(.tag)[idx]) {
        .root, .struct_decl => .{
            .token = self.nodes.items(.main_idx)[idx],
            .members = self.extra_data[lhs..rhs],
        },
        else => null,
    };
}

pub fn assembledVarDecl(self: *const Ast, node_idx: Node.Idx) ?Assembled.VarDecl {
    const idx = @intFromEnum(node_idx);
    const token = self.nodes.items(.main_idx)[idx];
    const type_expr_idx = self.nodes.items(.data)[idx].lhs;
    const type_expr = if (type_expr_idx != Node.null_node)
        self.extra_data[@intFromEnum(type_expr_idx)]
    else
        Node.null_node;
    const expr = self.nodes.items(.data)[idx].rhs;
    return switch (self.nodes.items(.tag)[idx]) {
        .var_decl => .{
            .token = token,
            .type_expr = type_expr,
            .expr = expr,
        },
        else => null,
    };
}

File: src/ast.rs
use crate::{builtin::Mutability, parser::TaraParser};
use anyhow::Result;
use codespan::Span;
use num_bigint::BigUint;
use std::fmt::Display;
use symbol_table::GlobalSymbol;

pub struct Node {
    pub span: Span,
    pub kind: NodeKind,
}

impl Node {
    pub fn new(kind: NodeKind, span: pest::Span) -> Self {
        return Self {
            span: Span::new(span.start() as u32, span.end() as u32),
            kind,
        };
    }
}

impl Display for Node {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind)?;
        return Ok(());
    }
}

pub enum NodeKind {
    StructDecl(StructInner),
    VarDecl(VarDecl),
    ModuleDecl(ModuleInner),
    LocalVarDecl(VarDecl),
    Or(BinOp),
    And(BinOp),
    Lt(BinOp),
    Gt(BinOp),
    Lte(BinOp),
    Gte(BinOp),
    Eq(BinOp),
    Neq(BinOp),
    BitAnd(BinOp),
    BitOr(BinOp),
    BitXor(BinOp),
    Add(BinOp),
    Sub(BinOp),
    Mul(BinOp),
    Div(BinOp),
    Access(BinOp),
    Call(Call),
    Negate(UnOp),
    Deref(UnOp),
    Return(UnOp),
    Identifier(GlobalSymbol),
    ReferenceTy(RefTy),
    PointerTy(RefTy),
    NumberLiteral(BigUint),
    SizedNumberLiteral(SizedNumberLiteral),
    IfExpr(IfExpr),
    SubroutineDecl(SubroutineDecl),
    StructInit(StructInit),
}

impl Display for NodeKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            NodeKind::StructDecl(struct_inner) => {
                f.write_fmt(format_args!("struct_decl({})", struct_inner))?
            }
            NodeKind::VarDecl(var_decl) => f.write_fmt(format_args!("var_decl({})", var_decl))?,
            NodeKind::ModuleDecl(mod_inner) => {
                f.write_fmt(format_args!("module_decl({})", mod_inner))?
            }
            NodeKind::LocalVarDecl(var_decl) => f.write_fmt(format_args!("local({})", var_decl))?,
            NodeKind::Or(expr) => f.write_fmt(format_args!("or({})", expr))?,
            NodeKind::And(expr) => f.write_fmt(format_args!("and({})", expr))?,
            NodeKind::Lt(expr) => f.write_fmt(format_args!("lt({})", expr))?,
            NodeKind::Gt(expr) => f.write_fmt(format_args!("gt({})", expr))?,
            NodeKind::Lte(expr) => f.write_fmt(format_args!("lte({})", expr))?,
            NodeKind::Gte(expr) => f.write_fmt(format_args!("gte({})", expr))?,
            NodeKind::Eq(expr) => f.write_fmt(format_args!("eq({})", expr))?,
            NodeKind::Neq(expr) => f.write_fmt(format_args!("neq({})", expr))?,
            NodeKind::BitAnd(expr) => f.write_fmt(format_args!("bit_and({})", expr))?,
            NodeKind::BitOr(expr) => f.write_fmt(format_args!("bit_or({})", expr))?,
            NodeKind::BitXor(expr) => f.write_fmt(format_args!("bit_xor({})", expr))?,
            NodeKind::Add(expr) => f.write_fmt(format_args!("add({})", expr))?,
            NodeKind::Sub(expr) => f.write_fmt(format_args!("sub({})", expr))?,
            NodeKind::Mul(expr) => f.write_fmt(format_args!("mul({})", expr))?,
            NodeKind::Div(expr) => f.write_fmt(format_args!("div({})", expr))?,
            NodeKind::Access(expr) => f.write_fmt(format_args!("access({})", expr))?,
            NodeKind::Call(expr) => f.write_fmt(format_args!("call({})", expr))?,
            NodeKind::Identifier(ident) => {
                f.write_fmt(format_args!("identifier(\"{}\")", ident.as_str()))?
            }
            NodeKind::Negate(expr) => f.write_fmt(format_args!("neg({})", expr))?,
            NodeKind::Deref(expr) => f.write_fmt(format_args!("deref({})", expr))?,
            NodeKind::Return(expr) => f.write_fmt(format_args!("return({})", expr))?,
            NodeKind::ReferenceTy(expr) => f.write_fmt(format_args!("reference_ty({})", expr))?,
            NodeKind::PointerTy(expr) => f.write_fmt(format_args!("pointer_ty({})", expr))?,
            NodeKind::NumberLiteral(num) => f.write_fmt(format_args!("number({})", num))?,
            NodeKind::SizedNumberLiteral(num) => {
                f.write_fmt(format_args!("sized_number({})", num))?
            }
            NodeKind::IfExpr(expr) => f.write_fmt(format_args!("if_expr({})", expr))?,
            NodeKind::SubroutineDecl(subroutine) => {
                f.write_fmt(format_args!("subroutine_decl({})", subroutine))?
            }
            NodeKind::StructInit(expr) => f.write_fmt(format_args!("struct_init({})", expr))?,
        }
        Ok(())
    }
}

pub struct StructInner {
    pub fields: Vec<TypedName>,
    pub members: Vec<Node>,
}

impl StructInner {
    pub fn new(fields: Vec<TypedName>, members: Vec<Node>) -> Self {
        return Self { fields, members };
    }
}

impl Display for StructInner {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("struct_inner(")?;
        for field in &self.fields {
            f.write_fmt(format_args!("({}, {}), ", field.name, *field.ty))?;
        }
        for member in &self.members {
            f.write_fmt(format_args!("({}), ", member))?;
        }
        f.write_str(")")?;
        Ok(())
    }
}

pub struct ModuleInner {
    pub fields: Vec<TypedName>,
    pub members: Vec<Node>,
}

impl ModuleInner {
    pub fn new(fields: Vec<TypedName>, members: Vec<Node>) -> Self {
        return Self { fields, members };
    }
}

impl Display for ModuleInner {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("module_inner(")?;
        for field in &self.fields {
            f.write_fmt(format_args!("({}, {}), ", field.name, *field.ty))?;
        }
        for member in &self.members {
            f.write_fmt(format_args!("({}), ", member))?;
        }
        f.write_str(")")?;
        Ok(())
    }
}

pub enum Publicity {
    Public,
    Private,
}

impl Display for Publicity {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Publicity::Public => "pub",
            Publicity::Private => "priv",
        };
        f.write_str(s)?;
        Ok(())
    }
}

pub struct VarDecl {
    pub publicity: Publicity,
    pub ident: GlobalSymbol,
    pub ty: Option<Box<Node>>,
    pub expr: Box<Node>,
}

impl VarDecl {
    pub fn new(publicity: Publicity, ident: GlobalSymbol, ty: Option<Node>, expr: Node) -> Self {
        return Self {
            publicity,
            ident,
            ty: ty.map(Box::new),
            expr: Box::new(expr),
        };
    }
}

impl Display for VarDecl {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.ty {
            Some(node) => f.write_fmt(format_args!(
                "{} \"{}\": {} = {}",
                self.publicity,
                self.ident.as_str(),
                *node,
                *self.expr
            ))?,
            None => f.write_fmt(format_args!(
                "{} \"{}\" = {}",
                self.publicity,
                self.ident.as_str(),
                *self.expr
            ))?,
        }
        Ok(())
    }
}

pub struct TypedName {
    pub ty: Box<Node>,
    pub name: GlobalSymbol,
}

impl TypedName {
    pub fn new(ty: Node, name: GlobalSymbol) -> Self {
        return Self {
            ty: Box::new(ty),
            name,
        };
    }
}

pub struct BinOp {
    pub lhs: Box<Node>,
    pub rhs: Box<Node>,
}

impl Display for BinOp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{}, {}", *self.lhs, *self.rhs))?;
        Ok(())
    }
}

pub struct UnOp {
    pub lhs: Box<Node>,
}

impl Display for UnOp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{}", *self.lhs))?;
        Ok(())
    }
}

pub struct Call {
    pub call: Box<Node>,
    pub args: Vec<Node>,
}

impl Display for Call {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{}, (", *self.call))?;
        for arg in &self.args {
            f.write_fmt(format_args!("{}", arg))?;
        }
        f.write_str(")")?;
        Ok(())
    }
}

pub struct SizedNumberLiteral {
    pub size: u16,
    pub literal: BigUint,
}

impl Display for SizedNumberLiteral {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{}'d{}", self.size, self.literal))?;
        Ok(())
    }
}

pub struct RefTy {
    pub mutability: Mutability,
    pub ty: Box<Node>,
}

impl RefTy {
    pub fn new(mutability: Mutability, ty: Node) -> Self {
        return Self {
            mutability,
            ty: Box::new(ty),
        };
    }
}

impl Display for RefTy {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{} {}", self.mutability, *self.ty))?;
        Ok(())
    }
}

pub struct IfExpr {
    pub cond: Box<Node>,
    pub body: Box<Node>,
    pub else_body: Box<Node>,
}

impl Display for IfExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!(
            "{}, {}, {}",
            *self.cond, *self.body, *self.else_body
        ))?;
        Ok(())
    }
}

pub struct SubroutineDecl {
    pub publicity: Publicity,
    pub ident: GlobalSymbol,
    pub params: Vec<TypedName>,
    pub return_type: Box<Node>,
    pub block: Vec<Node>,
}

impl SubroutineDecl {
    pub fn new(
        publicity: Publicity,
        ident: GlobalSymbol,
        params: Vec<TypedName>,
        return_type: Node,
        block: Vec<Node>,
    ) -> Self {
        return Self {
            publicity,
            ident,
            params,
            return_type: Box::new(return_type),
            block,
        };
    }
}

impl Display for SubroutineDecl {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{} {}(", self.publicity, self.ident))?;
        for param in &self.params {
            f.write_fmt(format_args!("({}, {}), ", param.name, *param.ty))?;
        }
        f.write_fmt(format_args!(") {} (", *self.return_type))?;
        for statement in &self.block {
            f.write_fmt(format_args!("({}), ", statement))?;
        }
        f.write_str(")")?;
        Ok(())
    }
}

pub struct StructInit {
    pub ty: Option<Box<Node>>,
    // Technically not a typed name, but a named value but whatever
    pub fields: Vec<TypedName>,
}

impl StructInit {
    pub fn new_with_ty(ty: Node, fields: Vec<TypedName>) -> Self {
        return Self {
            ty: Some(Box::new(ty)),
            fields,
        };
    }

    pub fn new_anon(fields: Vec<TypedName>) -> Self {
        return Self { ty: None, fields };
    }
}

impl Display for StructInit {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(expr) = &self.ty {
            f.write_fmt(format_args!("{}, (", expr))?;
        } else {
            f.write_str("anon, (")?;
        }
        for field in &self.fields {
            f.write_fmt(format_args!(".{} = {}, ", field.name, *field.ty))?;
        }
        f.write_str(")")?;
        Ok(())
    }
}

pub struct Ast {
    pub root: Node,
}

impl Ast {
    pub fn parse(source: &str) -> Result<Ast> {
        let root = TaraParser::parse_source(source)?;

        return Ok(Self { root });
    }
}

impl Display for Ast {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{}", self.root))
    }
}
